pkg/util/sketch/cms.go

```diff
@@ -0,0 +1,117 @@
+package sketch
+
+import (
+	"fmt"
+	"math"
+)
+
+type CountMinSketch struct {
+	depth, width int
+	counters     [][]uint32
+}
+
+// NewCountMinSketch creates a new CMS for a given width and depth.
+func NewCountMinSketch(w, d int) (*CountMinSketch, error) {
+	if d < 1 || w < 1 {
+		return nil, fmt.Errorf("sketch dimensions must be positive, w: %d, d: %d", w, d)
+	}
+
+	return &CountMinSketch{
+		depth:    d,
+		width:    w,
+		counters: make2dslice(w, d),
+	}, nil
+}
+
+func make2dslice(col, row int) [][]uint32 {
+	ret := make([][]uint32, row)
+	for i := range ret {
+		ret[i] = make([]uint32, col)
+	}
+	return ret
+}
+
+func (s *CountMinSketch) getPos(h1, h2 uint32, row int) uint32 {
+	pos := (h1 + uint32(row)*h2) % uint32(s.width)
+	return pos
+}
+
+// Add 'count' occurences of the given input.
+func (s *CountMinSketch) Add(event string, count int) {
```

Comment:
I wonder if we should make the event type generic. Maybe that's something for the next iteration.

Reply:
I'm not sure we do, is there a use case you're thinking of? 

The string of labels to identify each thing we're doing a topk on are what we'll want to use to insert things into the sketch. This could be the entire labelset for a stream, or partial, or with additional labels from line parsing. So we can't just use for example the already computed stream fingerprint.

Reply:
I think the sample stream would give us the fingerprints but it's not urgent now. Let's roll with strings.